### ISSUE TITLE
E0569 is not detected for methods implemented for mutable reference

### ISSUE THREAD

**Issue by Undin** [intellij-rust/intellij-rust#9448]
## Environment

* **IntelliJ Rust plugin version:** 0.4.180.4912-222-nightly
* **Rust toolchain version:** 1.64.0 (a55dd71d5 2022-09-19) x86_64-apple-darwin
* **IDE name and version:** IntelliJ IDEA 2022.2 Ultimate Edition (IU-222.3345.118)
* **Operating system:** macOS 12.5.1
* **Macro expansion:** enabled
* **Additional experimental features:** org.rust.cargo.features.settings.gutter, org.rust.cargo.emulate.terminal, org.rust.cargo.evaluate.build.scripts, org.rust.macros.proc

## Problem description
The plugin doesn't detect [E0569]() when a method implemented for a mutable reference type.

```rust
trait Foo {
    fn foo(self) where Self: Sized {}
}

struct S;

impl<'a> Foo for &'a mut S {}

fn main() {
    let s = S;
    s.foo(); // E0569 is not detected
}
```

<img width="395" alt="image" src="https://user-images.githubusercontent.com/2539310/193560320-619749fb-1808-4ee3-86d7-a075fa4171c6.png">

At the same time, if we move mutability restriction to method itself, the error will be detected by the plugin

```rust
trait Foo {
    fn foo(&mut self) where Self: Sized {}
}

struct S;

impl<'a> Foo for S {}

fn main() {
    let s = S;
    s.foo(); // E0569 is detected
}
```
<img width="499" alt="image" src="https://user-images.githubusercontent.com/2539310/193559792-ded624c8-bee4-4e88-b369-de6f74f353ce.png">
